cmake_library_suffix_is_bc

juj opened the pull request:
Add CMake command line options `-DEMSCRIPTEN_CMAKE_STATIC_LIBRARY_SUFFIX_IS_BC=ON/OFF` and `-DEMSCRIPTEN_CMAKE_SHARED_LIBRARY_SUFFIX_IS_BC=ON/OFF` for configuring target library suffixes on the command line.

kripken commented:
Interesting, what is the motivation here?

Does this change the existing behavior? If so, tests should need updating (or new tests if this wasn't covered)?

dschuff commented:
Not sure about the context of this either but I would definitely be in favor of moving toward always using `.a` instead of `.bc` for libraries, which is much more conventional.

dschuff commented:
(Also it will make LTO faster)

juj commented:
Rewrote this to work better:
  - Renamed the option to `-DEMSCRIPTEN_GENERATE_BITCODE_STATIC_LIBRARIES=ON` to better reflect the intent.
  - Removed the same option for shared libraries. We probably want to do something there as well, but I don't know what until we get real dynamic linking in wasm available. (For now, all CMake users should just generate static libraries, shared libs are just disguised static libs)
  - Added a test

juj commented:
>Interesting, what is the motivation here?

The origin of this is that I'm troubleshooting a codebase which has trouble linking their project together, and the culprit was that they were overriding the static library suffix from `.a` to `.bc` in CMake, with the expectation that that would have changed from Unix archives to LLVM bitcodes, but it by itself didnt't do anything but just renamed the suffix. This adds a proper option to be able to do that.

> Does this change the existing behavior? If so, tests should need updating (or new tests if this wasn't covered)?

This does not change existing behavior. The default behavior still is, like it used to be (and it is intended to be in the future) to output Unix .a archives from CMake static library targets. This is in line with how CMake works natively. Added a test for this now (there can't be too much testing for CMake stuff, it tends to be an extremely brittle piece of technology)

juj commented:
> Not sure about the context of this either but I would definitely be in favor of moving toward always using .a instead of .bc for libraries, which is much more conventional.

Using .a is the default for Emscripten CMake support, has always been. Perhaps it might not be in `./configure` emulation land though (not too familiar with that machinery).

This PR adds support to cleanly target LLVM bitcode where it makes sense to do so. The new feature is intended to be an explicit opt-in option, which will not become a default or anything like that.

kripken commented on cmake/Modules/Platform/Emscripten.cmake:
```diff
@@ -166,12 +166,24 @@ set(CMAKE_SYSTEM_INCLUDE_PATH "${EMSCRIPTEN_ROOT_PATH}/system/include")
 # but if these are adjusted, ${CMAKE_ROOT}/Modules/CheckIncludeFile.cmake will fail, since it depends on being able to compile output files with predefined names.
 #SET(CMAKE_LINK_LIBRARY_SUFFIX "")
 #SET(CMAKE_STATIC_LIBRARY_PREFIX "")
-#SET(CMAKE_STATIC_LIBRARY_SUFFIX ".bc")
```
granted I know nothing about cmake ;) but to me this looks like you are removing a line that defaulted the suffix to `.bc`, and later below do that only if an option is set. So it appears to change the default behavior? I'm sure i'm missing something...

juj replied:
Oh, `#` is a comment sign in CMake, these lines had been commented out. (but left for later reference) The default option has been CMake's default, which is `.a`.

kripken replied:
Oh, heh, good to know!

juj commented:
Sounds like good to merge, so pushing this in.
